### ISSUE TITLE
Existing commits are checked at branch creation, but they shouldn't be checked

### ISSUE THREAD

**Issue by GyulaJuhasz** [mohamicorp/yet-another-commit-checker#150]
On the plugin's description, this is stated: "Branch friendly! Only new commits are checked. Commits that already exist in the repository will be skipped."

However, I cannot create and push a new branch from a tag if the tag has commits in its history that are not considered valid.

Here is what I do:

```
git checkout -b newBranch tag-with-invalid-commits.
git push -u origin newBranch
```

**Expected result:** the push succeeds, since I am not pushing  any new commits, I'm just trying to push a branch ref to the remote.

**Actual result:** the push is rejected, because the plugin compares the old commits with my current name and email, which will fail for two obvious reasons:
- there are commits committed by others
- there are commits that did not follow the conventions.

The workaround I use now is to create the branch by using the BitBucket UI. However, I am certain that this should work using git commands too.

Another interesting fact is that after I successfully created a branch from a tag using the UI, I can create and push branches from the same tag without any issues.

Please investigate this.

Thank You in advance.

Update: it seems that the problem arises if the tag is not reachable from any branch. Based on this assumption, here are a few detailed replication steps:

1. Create a new branch: `git checkout -b newBranch`
2. Turn off commit checker.
3. Create commits that would violate the rules (wrong committer name, etc.)
4. Push the commits to the remote: `git push -u origin newBranch`
5. Create a tag: `git tag -a newTag -m "newTag"`
6. Push the tag: `git push --tags`
7. Checkout any other branch: `git checkout master`
8. Remove the local branch: `git branch -D newBranch`
9. Remove the branch from the remote: `git push origin :refs/heads/newBranch`
10. Turn on the commit checker.

After these steps, there are invalid commits reachable from `newTag`, but not reachable from any branches, and the initial replication steps should show the problem:

```
git checkout -b brandNewBranch newTag
git push -u origin brandNewBranch
```

The push should fail because of the commits that already exist on the remote.

**Comment by sford** [comment]
What version of BitBucket, YACC, and git are you using?

**Comment by GyulaJuhasz** [comment]
Hi,

Sorry for the late reponse. The versions:

BitBucket: v4.5.2
Git: 2.12.2.windows.2
YACC: v1.15

Thank You for investigating this.

**Comment by sford** [comment]
Is that the version of git you are using or BBS is using (or both)?

**Comment by GyulaJuhasz** [comment]
It's the version I used by me locally to create new branches.

The same error occurs when I try to create and push branches from IDEs (IntelliJ IDEA/Android Studio).

The only possible way that's left to create branches is by using the web-based Bitbucket UI.

(I don't know what Git version Bitbucket is using.)

**Comment by itay** [comment]
@sford if I had to guess, it's because this commit might only be referred to by a tag and may not be reachable by any named branch. If that's the case, I think the input you pass into `git rev-list` to see the list of commits that are reachable might not be seeing that commit, and thus you run the YACC stuff on it again.

**Comment by GyulaJuhasz** [comment]
@itay:

> if I had to guess, it's because this commit might only be referred to by a tag and may not be reachable by any named branch.


This exactly is the situation here. I checked the code, and it seems like it is prepared for this case. ("git rev-list <hash> --not --all") The --all includes every ref, including the tags too.

I forked the repository, and created a Pull Request. I hope the PR can help in solving this.

**Comment by sford** [comment]
@itay That is really good guess... I was thinking along those lines too. In fact, YACC <1.13 worked by checking named branches, which would certainly miss commits not reachable by a branch. As @GyulaJuhasz mentioned, newer versions use `rev-list` which I would expect handle this.

**Comment by GyulaJuhasz** [comment]
Can you please give me an update about this issue?

I see that 1.16 has been released. Does it work with BitBucket 4.5.2? If yes, then I will test the new version.

**Comment by RobSmyth** [comment]
Sorry, I no longer work at the company where I was seeing this issue. So I can not provide any further update. The issue may be closed.

**Comment by sford** [comment]
@GyulaJuhasz I just released YACC 1.17 which removes the git rev-list calls in favor of new BBS 5 hook APIs to obtain commit information. In other words, it is now BBS's responsibility to determine which commits are new. This may have an impact (hopefully a fix) on what you are running into.

If you are still using BBS 4.5.2, you could try upgrading to 5+ to see if it fixes your problem.
